src/main/java/de/tum/in/www1/artemis/service/ModelingSubmissionService.java

```diff
@@ -111,32 +111,44 @@ public ModelingSubmission getLockedModelingSubmissionWithoutResult(ModelingExerc
     }
 
     /**
-     * Given an exercise, find a modeling submission for that exercise which still doesn't have any result. If the diagram type is supported by Compass we get the next optimal
+     * Given an exercise, find a modeling submission for that exercise which still doesn't have a manual result. If the diagram type is supported by Compass we get the next optimal
      * submission from Compass, i.e. the submission for which an assessment means the most knowledge gain for the automatic assessment mechanism. If it's not supported by Compass
-     * we just get a random submission without assessment.
+     * we just get a random submission without assessment. If there is no submission without manual result we return an empty optional. Note, that we cannot use a readonly
+     * transaction here as it is making problems when initially loading the calculation engine and assessing all submissions automatically: we would get an sql exception
+     * "Connection is read-only" from hibernate when saving the result in CompassService#assessAutomatically.
      *
      * @param modelingExercise the modeling exercise for which we want to get a modeling submission without result
      * @return a modeling submission without any result
      */
-    @Transactional(readOnly = true)
+    @Transactional
     public Optional<ModelingSubmission> getModelingSubmissionWithoutResult(ModelingExercise modelingExercise) {
-        // ask Compass for optimal submission to assess if diagram type is supported
+        // if the diagram type is supported by Compass, ask Compass for optimal (i.e. most knowledge gain for automatic assessments) submissions to assess next
         if (compassService.isSupported(modelingExercise.getDiagramType())) {
-            Set<Long> optimalModelSubmissions = compassService.getModelsWaitingForAssessment(modelingExercise.getId());
-            if (!optimalModelSubmissions.isEmpty()) {
-                // TODO CZ: think about how to handle canceled assessments with Compass as I do not want to receive the same submission again, if I canceled the assessment
-                return modelingSubmissionRepository.findById(optimalModelSubmissions.iterator().next());
+            List<Long> modelsWaitingForAssessment = compassService.getModelsWaitingForAssessment(modelingExercise.getId());
+
+            // shuffle the model list to prevent that the user gets the same submission again after canceling an assessment
+            Collections.shuffle(modelsWaitingForAssessment);
+
+            for (Long submissionId : modelsWaitingForAssessment) {
```

Comment:
As far as I understand it, this could result in a lot of queries on the sql database.

It might be possible to do this whole operation in sql (2 queries):
1. Get ids of modelsWaitingForAssessment, shuffle them
2. Query with the list of ids the models joined with the submissions (right join), limit 1.

Reply:
Thanks for the comment. But it cannot result in many sql queries for two reasons:
1) The `modelsWaitingForAssessment` are limited to 10 models in the `CompassService`.
2) The method should return in the first iteration of the for-loop as there should only be existing models in `modelsWaitingForAssessment`. I just added the for-loop for the case that something went wrong with the `modelsWaitingForAssessment` list, which should not happen. I had this case once after I manually deleted participations (together with the submissions) in an exercise and did not restart the server, so the id of the deleted submission was still in the list and it did not find a submission in the first iteration. Normally, existing participations should not get deleted though.

I think we can keep it like this or did I miss anything?

Reply:
If it is only 10 models I think it's ok to keep it this way :+1: